### DISCUSSION TITLE
How do you "factory" reset? (and a few other questions)

### DISCUSSION THREAD

**Discussion by Sarkasaa** [nextcloud/recognize#626] [Q&A] [answered]
This has probably been asked before, but I can't seem to find a definitive answer on how to perform a full reset of the application (as in, how do I make it, so it's like I installed the app for the first time), including all tags, faces and so on.

I was organizing the recognized faces (I only use the face recognition) and because I didn't understand the "Merge with different Person" command I merged two people instead of just the photos I selected. My thought process was, that if I remove the whole person, the images will be reclassified on the next run, but after doing some research here, I wasn't so sure anymore. So I did what I do best: I ran all the resetting commands available in the Admin Settings UI and with "occ recognize" hoping to do a full rerun. Since nothing was happening, I went back to GitHub and did some more reading and just got more confused.

So what I'm asking myself is:

1. What happens to "removed person" photos? Do they get reclassified? Or are they permanently removed? And if they are permanently removed, how do I re-add them to the queue?
2. What happens to "remove photos from person" photos? Do they get reclassified? Or are they permanently removed? And if they are permanently removed, how do I re-add them to the queue?
3. How do I go about resetting the database safely? Non-named classified people have a number, that is supposed to start at 1, but after all the stuff I did to reset, the freshly classified photos have a much higher number, indicating to me, that the previous database entries still exist.

**Comment by zyberwoof** [comment]
I would also like this kind of info.  I've had several issues with Recognize in the few months that I've been trying it.  Some or all of these issues could be due to me miss-configuring things, or Nextcloud bugs.  So I'm not pointing any fingers at anyone.  But as everyone knows, "rebooting" is helpful.  Both for fixing reoccurring issues, and for troubleshooting root causes.  

It would be helpful to have a little more clarity on the occ Recognize commands available to users now.  And if a user would like to tinker, what's the proper way of fully resetting or removing the Recognize app?  The goal would be to let users do some trial-and-error without fear of corrupting the rest of their Nextcloud installation.

**Reply by marcelklehr** [reply]
I added some notes in my other comment here, the admin UI should already contain some helpful notes about the various commands, though.

**Comment by marcelklehr** [comment, marked as answer]
There are two stacks of face detections other than the face clusters you can see in the photos app:

 * The first is "fresh" face detections, which haven't been run through clustering yet.
 * The second is face detections which have been run through clustering but weren't assigned to any clusters, these we call "rejected" face detections.

When you delete a person / face cluster from your photos, those face detections end up on the rejected face detections stack. When you remove a face detection from a person / face cluster, it lands on the "fresh" detections pile with a small annotation to make sure it doesn't end up with the same wrong person again.

When face clustering runs, it will take as many detections from the "fresh" pile as it can, and mix in a smaller, random selection from the "rejected" pile to try and get those detections assigned to a cluster.

> after all the stuff I did to reset, the freshly classified photos have a much higher number

The numbers have no meaning. Feel free to ignore them.

# How do I go about resetting the database safely?

## Faces

 * `occ recognize:reset-face-clusters` will remove all people / face clusters from all users and put all face detections on the "fresh" pile.
 * `occ recognize:cluster-faces -b 10000` will then run the clustering algorithm for a batch of 10000 detections (you can also leave out the -b option, but you may run out of RAM).
 * `occ recognize:reset-faces` will remove all face detections from your database. If you want to re-add them, you need to run either `occ recognize:classify` or `occ recognize:recrawl`.

## Object detection, Landmarks detection, music and video recognition

 * `occ recognize:reset-tags` will remove all tag assignments from your photos. If you want to re-add those tags, you need to run either `occ recognize:classify` or `occ recognize:recrawl`.
 * (optional) `occ recognize:remove-legacy-tags` will remove tag assignments from your photos that have been created by earlier recognize versions. Use this if the first command doesn't remove everything.

## Background jobs

 * `occ recognize:clear-background-jobs` will remove any active background jobs that stem from scanning your files and will clear all background job queues in the database. If you want to scan all files again, you need to run either `occ recognize:classify` or `occ recognize:recrawl`.

## `occ recognize:classify` vs `occ recognize:recrawl`

### classify
The classify command is synchronous and runs directly in your terminal, possibly for hours. It goes through all user's files and runs the classifiers directly as subprocesses of the terminal command. When starting out with recognize, it will automatically run a background scan. You can stop it with `occ recognize:clear-background-jobs` and then run `occ recognize:classify` to do the initial scan manually on the terminal instead. After running `occ recognize:classify`  whenever new files are added, recognize will still scan those files using the background mechanism that is also behind `occ recognize:recrawl`.

### Recrawl
The recrawl command clears all queues and background jobs and schedules new background jobs to scan all files in background processes. This is slower than `classify` but also less resource intensive.

**Comment by palves** [comment]
Does "occ recognize:classify" reclassify all files, even if they have been classified before, or does it skips files that have already been classified previously?
